firedrake/output.py

```diff
@@ -651,3 +754,381 @@ def write(self, *functions, **kwargs):
+        mesh.init()
+        topology_dm = mesh.topology.topology_dm
+        topology_name = topology_dm.getName()
+        if topology_name in self.h5pyfile.require_group(self.path_to_topologies()):
+            # Check if the global number of DMPlex points and
+            # the global sum of DMPlex cone sizes are consistent.
+            order_array_size, ornt_array_size = dmcommon.compute_point_cone_global_sizes(mesh)
+            # TODO: Fix this once PETSc default path to topology is fixed.
+            path = "/topology"
+            order_array_size1 = self.h5pyfile[path]["order"].size
+            ornt_array_size1 = self.h5pyfile[path]["orientation"].size
+            filename = self.viewer.getFileName()
+            if order_array_size1 != order_array_size:
+                raise ValueError(f"Mesh named {topology_name} already exists in {filename}, but the global number of DMPlex points is inconsistent: {order_array_size1} ({filename}) != {order_array_size} ({topology_name})")
+            if ornt_array_size1 != ornt_array_size:
+                raise ValueError(f"Mesh named {topology_name} already exists in {filename}, but the global sum of all DMPlex cone sizes is inconsistent: {ornt_array_size1} ({filename}) != {ornt_array_size} ({topology_name})")
+            return
+        self.viewer.pushFormat(format=ViewerHDF5.Format.HDF5_PETSC)
+        topology_dm.topologyView(viewer=self.viewer)
+        topology_dm.labelsView(viewer=self.viewer)
+        topology_dm.coordinatesView(viewer=self.viewer)
+        self.viewer.popFormat()
+
+    def save_function_space(self, V):
+        if isinstance(V.topological, impl.MixedFunctionSpace):
+            base_path = self.path_to_mixed_function_space(V.name)
+            self.h5pyfile.require_group(base_path)
+            self.set_attr(base_path, PREFIX + "_num_sub_spaces", V.num_sub_spaces())
+            for i, Vsub in enumerate(V):
+                path = os.path.join(base_path, str(i))
+                self.h5pyfile.require_group(path)
+                self.set_attr(path, PREFIX + "_mesh", Vsub.mesh().name)
+                self.set_attr(path, PREFIX + "_function_space", Vsub.name)
+                self.save_function_space(Vsub)
+        else:
+            # Save mesh
+            self.save_mesh(V.mesh())
+            # Save function space
+            tV = V.topological
+            topology_dm = tV.mesh().topology_dm
+            dm = get_sectiondm_for_checkpointing(tV)
+            topology_name = topology_dm.getName()
+            dm_name = dm.getName()
+            if tV.name not in self.h5pyfile.require_group(self.path_to_function_spaces(topology_name, dm_name)):
+                # Save section
+                topology_dm.sectionView(self.viewer, dm)
+                # Save UFL element
+                element = tV.ufl_element()
+                element_b = pickle.dumps(element)
+                path = self.path_to_function_space(topology_name, dm_name, tV.name)
```

Comment:
It seems that two functionspaces with the same data layout but different UFL elements (and the same name, which is often `None` because that's the default) will map to the same place here. e.g. a `FiniteElement("DG", triangle, 0)` and a `FiniteElement("B", triangle, 3)` have the same entity dofs, but vastly different meanings.

Reply:
Yes. This is related to other comments of yours, but, currently, we assume that users name `FunctionSpace`s when they want to save `Function`s defined on them. Maybe we should automatically construct `FunctionSpace` names from UFL element somehow?

Reply:
People aren't thinking about saving function spaces, they're thinking about saving functions, so I don't think we should require functionspaces to have names. Can we just give them generated (unique per section) names in the file and save an attribute on the section that maps elements to location (e.g. via a pickled dict).

Reply:
Yes. I think that makes sense. I'll work on that. Thanks Lawrence!

Reply:
Updated. I think it looks much cleaner now.